File: PalindromeLinkedList/PalindromeLinkedList.cpp
/**
* Definition for singly-linked list.
*/
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
	bool isPalindrome(ListNode* head) {
		if (head == nullptr || head->next == nullptr)
			return true;
		ListNode* slow = head, *fast = head;
		while (fast->next != nullptr && fast->next->next != nullptr) {
			slow = slow->next;
			fast = fast->next->next;
		}
		slow->next = reverseList(slow->next);
		slow = slow->next;
		while (slow != nullptr) {
			if (head->val != slow->val)
				return false;
			head = head->next;
			slow = slow->next;
		}
		return true;
	}

	ListNode* reverseList(ListNode* head) {
		ListNode *pre = nullptr;
		ListNode *cur = head;
		while (cur != nullptr) {
			ListNode* next = cur->next;
			cur->next = pre;
			pre = cur;
			cur = next;
		}
		return pre;
	}
};

File: PalindromeLinkedList/PalindromeLinkedList.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head == None or head.next == None:
            return True
        # find the middle of the LinkedList
        fast = head
        slow = head
        while fast != None and fast.next != None:
            fast = fast.next.next
            slow = slow.next
        
        if fast != None:
            slow.next = self.Reverse(slow.next)
            slow = slow.next
        else:
            slow = self.Reverse(slow)
        
        while slow != None:
            if slow.val != head.val:
                return False
            slow = slow.next
            head = head.next
        return True
    
    
    def Reverse(self, head):
        if head is None or head.next is None:
            return head
        prev = None
        while head != None:
            curr = head
            head = head.next
            curr.next = prev
            prev = curr
        return prev
        